pkg/ddevapp/providerDrudS3.go

```diff
@@ -0,0 +1,392 @@
+	fmt.Print(environmentPrompt + ": ")
+	envName := util.GetInput(p.EnvironmentName)
+
+	// TODO: Get this to validate environment correctly
+
+	//if !ok {
+	//	return fmt.Errorf("could not find an environment named '%s'", envName)
+	//}
+	p.SetSiteNameAndEnv(envName)
+	return nil
+}
+
+// Write the DrudS3 provider configuration to a spcified location on disk.
+func (p *DrudS3Provider) Write(configPath string) error {
+	err := PrepDdevDirectory(filepath.Dir(configPath))
+	if err != nil {
+		return err
+	}
+
+	cfgbytes, err := yaml.Marshal(p)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(configPath, cfgbytes, 0644)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Read DrudS3 provider configuration from a specified location on disk.
+func (p *DrudS3Provider) Read(configPath string) error {
+	source, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return err
+	}
+
+	// Read config values from file.
+	err = yaml.Unmarshal(source, p)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetEnvironments will return a list of environments for the currently configured upstream DrudS3 site.
+func (p *DrudS3Provider) GetEnvironments() (map[string]bool, error) {
+	environments, err := findDrudS3Project(p.app.Name)
+	if err != nil {
+		return nil, err
+	}
+	return environments, nil
+}
+
+// Validate ensures that the current configuration is valid (i.e. the configured DrudS3 site/environment exists)
+func (p *DrudS3Provider) Validate() error {
+	return p.environmentExists()
+}
+
+// environmentExists ensures the currently configured DrudS3 site & environment exists.
+func (p *DrudS3Provider) environmentExists() error {
+	environments, err := p.GetEnvironments()
+	if err != nil {
+		return err
+	}
+	if _, ok := environments[p.EnvironmentName]; !ok {
+		return fmt.Errorf("could not find an environment with backups named '%s'", p.EnvironmentName)
+	}
+
+	return nil
+}
+
+// findDrudS3Project ensures the DrudS3 site specified by name exists, and the current user has access to it.
+// It returns an array of environment names and err
+func findDrudS3Project(project string) (map[string]bool, error) {
+	_, client, err := getDrudS3Session()
+	if err != nil {
+		return nil, err
+	}
+	// Get a list of all projects the current user has access to.
+	projectMap, err := getDrudS3Projects(client, DrudS3BucketName)
+	if err != nil {
+		return nil, err
+	}
+
+	projectWithEnvs := projectMap[project]
+
+	// Get a list of environments for a given project.
+	return projectWithEnvs, nil
+}
+
+// getDrudS3Session loads the DrudS3 API config from disk and returns a DrudS3 session struct.
+func getDrudS3Session() (*session.Session, *s3.S3, error) {
+	//ddevDir := util.GetGlobalDdevDir()
+
+	// TODO: Move S3 auth into global config?
+	//sessionLocation := filepath.Join(ddevDir, "DrudS3config.json")
+
+	// This is currently depending on auth in ~/.aws/credentials
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String("us-east-1")},
```

Comment:
Good point - It actually doesn't matter for this kind of thing if I understand right, it's just declaring who we are using to *query*, it would be important if we were creating buckets or stuff like that.